File: web/src/links.ts
/**
 * The link a result points at, and where on the page to look.
 *
 * The other half of `src/parts_index/core/links.py`. The export writes the link Chrome's way, because
 * only one form can be written to a file; this side, which is the only place that knows which viewer is
 * about to open it, flips it where Firefox needs the mirror image.
 *
 *     Chrome / PDFium   top is measured DOWNWARDS from the top of the page (Adobe's specification)
 *     Firefox / pdf.js  top is a PDF user-space y, measured UPWARDS from the bottom
 *
 * The page height rides along as `&h=`, a key neither viewer knows and both ignore, so the flip is
 * `h - top` and needs nothing the link does not already carry.
 *
 * `tests/fixtures/links_cases.json` is read by this file's test and by the Python one. Two
 * implementations of a single URL must not drift apart; add a case there before changing either.
 */

export const ZOOM = 200
/** Start the view slightly left of and above the label, so it is not pressed against the edge. */
export const MARGIN = 0.06

export interface LinkDoc {
  public_url?: string | null
  page_url_tpl?: string | null
  page_offset?: number | null
  w_pt?: number | null
  h_pt?: number | null
}

/** Fill one of the four templates in use: none, `#page=`, archive.org's leaf counter, or a `?q=` search. */
export function pageUrl(doc: LinkDoc, n: number, q = ''): string {
  const tpl = doc.page_url_tpl
  const url = doc.public_url ?? ''
  if (!tpl) return url
  const leaf = n - 1 + (doc.page_offset ?? 0)
  return tpl
    .replace('{url}', url)
    .replace('{n}', String(n))
    .replace('{leaf}', String(leaf))
    .replace('{q}', q)
}

/** …and, where the viewer can be told, the first box the part was read in. */
export function partUrl(doc: LinkDoc, n: number, boxes?: number[][] | string | null, q = ''): string {
  const url = pageUrl(doc, n, q)
  if (!url.includes('#page=') || !boxes) return url
  let box: number[] | undefined
  try {
    const parsed = typeof boxes === 'string' ? JSON.parse(boxes) : boxes
    box = parsed?.[0]
  } catch {
    return url
  }
  const { w_pt: w, h_pt: h } = doc
  const x0 = box?.[0]
  const y0 = box?.[1]
  if (x0 === undefined || y0 === undefined || !w || !h) return url
  const left = Math.max(0, x0 / 1000 - MARGIN) * w
  const top = Math.max(0, y0 / 1000 - MARGIN) * h // downwards from the top: Chrome's reading
  return `${url}&zoom=${ZOOM},${Math.round(left)},${Math.round(top)}&h=${Math.round(h)}`
}

/**
 * Turn a link written for Chrome into the same view in Firefox.
 *
 * A link with no `zoom` has nothing to flip, and one with no `&h=` cannot be flipped at all — the height
 * is what the mirror is taken about — so both are returned untouched rather than guessed at.
 */
export function forFirefox(url: string): string {
  const m = url.match(/&zoom=(\d+),(-?[\d.]+),(-?[\d.]+)&h=([\d.]+)/)
  if (!m) return url
  const [, zoom, left, top, h] = m
  const flipped = Math.max(0, Math.round(Number(h) - Number(top)))
  return url.replace(m[0], `&zoom=${zoom},${left},${flipped}`)
}

/** True when this browser measures `top` from the bottom of the page. */
export function needsFlip(ua: string = navigator.userAgent): boolean {
  return /firefox|seamonkey/i.test(ua)
}

/** The link to hand this visitor. */
export function forViewer(url: string, ua?: string): string {
  return needsFlip(ua) ? forFirefox(url) : url
}

File: src/parts_index/core/links.py
"""The link a result points at, and where on the page to look.

On a scanned sheet of a 1959 amplifier, ctrl+F finds nothing: the text is a picture. So a result does not
just link to the document, it links to the page and asks the viewer to show the part:

    <pdf>#page=2&zoom=200,51,519&h=617

`zoom=scale,left,top` is the PDF Open Parameters way of asking for a position, and it is the only one
both major viewers honour — Chrome ignores `view=FitR` (measured 2026-09-20 with the Bassman 5F6-A sheet
in Chromium 153 and Firefox 155). The catch is that they disagree about `top`:

    Chrome / PDFium   measured DOWNWARDS from the top of the page, which is Adobe's specification
    Firefox / pdf.js  a PDF user-space y, measured UPWARDS from the bottom

They are mirror images, so **one URL cannot satisfy both**. The link is written Chrome's way and carries
the page height as `&h=`, a key neither viewer knows and both ignore (measured: the rendering is
byte-identical with and without it), so the page offering the link can flip it where it knows better:

    top_firefox = h - top_chrome

That flip happens in the browser, which is the only place that knows which viewer is about to open the
link. `web/src/links.ts` is the other half of this module, and `tests/fixtures/links_cases.json` is
checked by both, because two implementations of one URL must not drift apart.

The box comes from the OCR, in thousandths of the page, and the page size in PDF points says how big a
thousandth is. Without a size — an archive.org reader link, a bare image, a page read before sizes were
recorded — the plain page link is returned rather than a guess.
"""
from __future__ import annotations

import json

ZOOM = 200          # a part label is readable at 200 % and a good part of the sheet is still in view
MARGIN = 0.06       # start the view slightly left of and above the label, so it is not against the edge


def page_url(doc: dict, n: int, q: str = "") -> str:
    """The public link to page `n` of this document.

    Four shapes are in use: no template at all (the document is one page, or a link to a landing page),
    `#page=` for a PDF, archive.org's reader, which counts leaves from zero, and a `?q=` search for a
    book whose reader has no page anchor.
    """
    tpl = doc.get("page_url_tpl")
    if not tpl:
        return doc.get("public_url") or ""
    return tpl.format(url=doc.get("public_url") or "", n=n,
                      leaf=n - 1 + (doc.get("page_offset") or 0), q=q)


def part_url(doc: dict, n: int, boxes=None, q: str = "") -> str:
    """…and, where the viewer can be told, the first box the part was read in."""
    url = page_url(doc, n, q)
    if "#page=" not in url or not boxes:
        return url
    try:
        x0, y0 = (json.loads(boxes)[0] if isinstance(boxes, str) else boxes[0])[:2]
        w_pt, h_pt = doc["w_pt"], doc["h_pt"]
    except (KeyError, IndexError, TypeError, ValueError):
        return url
    if not w_pt or not h_pt:
        return url
    left = max(0.0, x0 / 1000 - MARGIN) * w_pt
    top = max(0.0, y0 / 1000 - MARGIN) * h_pt          # downwards from the top: Chrome's reading
    return f"{url}&zoom={ZOOM},{left:.0f},{top:.0f}&h={h_pt:.0f}"
